api/src/controllers/zot4planimport.ts

```diff
@@ -0,0 +1,137 @@
+/**
+ @module Zot4PlanImportRoute
+*/
+
+import { z } from 'zod';
+import { publicProcedure, router } from '../helpers/trpc';
+import { SavedRoadmap, SavedPlannerQuarterData, QuarterName } from '@peterportal/types';
+
+type Zot4PlanSchedule = {
+  years: string[][][];
+  selectedPrograms: {
+    value: number;
+    label: string;
+    is_major: boolean;
+  }[][];
+  addedCourses: [];
+  courses: [];
+  apExam: {
+    id: number;
+    name: string;
+    score: number;
+    courses: string[];
+    GE: [];
+    units: number;
+  }[];
+};
+
+/**
+ * Get a JSON schedule from Zot4Plan by name; return an empty {} if it does not exist
+ */
+const getFromZot4Plan = async (scheduleName: string) => {
+  let res = {};
+  await fetch('https://api.zot4plan.com/api/loadSchedule/' + scheduleName, {
+    method: 'PUT',
+  })
+    .then((response) => {
+      if (!response.ok) {
+        return {};
+      }
+      return response.json();
+    })
+    .then((json) => {
+      res = json;
+    });
+  return res as Zot4PlanSchedule;
+};
+
+const zot4PlanImportRouter = router({
+  /**
+   * Get a roadmap formatted for PeterPortal based on a Zot4Plan schedule by name
+   * and labeled with years based on the current year and the student's year
+   */
+  getScheduleFormatted: publicProcedure
+    .input(z.object({ scheduleName: z.string(), studentYear: z.string() }))
+    .query(async ({ input }) => {
+      try {
+        // Get the raw schedule data
+        const originalScheduleRaw = await getFromZot4Plan(input.scheduleName);
+        if (Object.keys(originalScheduleRaw).length === 0) {
+          // Failed
+          return { planners: [], transfers: [] } as SavedRoadmap;
+        }
+
+        // Convert it to the PeterPortal roadmap format
+        const converted: SavedRoadmap = {
```

Comment:
This function is about getting the schedule, so any logic that doesn't fit the function name, i.e. converting a roadmap, should be in its own helper function (or even helper file).